hw/drivers/sensors/lis2dw12/src/lis2dw12.c

```diff
@@ -1863,60 +1915,92 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
 
-    rc = lis2dw12_get_fs(itf, &fs);
+    rc = lis2dw12_readlen(itf, LIS2DW12_REG_CTRL_REG1, prev_config, 6);
     if (rc) {
         return rc;
     }
-
-    /* ensure self test mode is disabled */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_DISABLE);
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG2, &st_config[1], 5);
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG1, st_config, 1);
     if (rc) {
         return rc;
     }
 
-    os_time_delay(OS_TICKS_PER_SEC / 10);
-
-    /* take base reading */
-    rc = lis2dw12_get_data(itf, fs, &(base[0]), &(base[1]), &(base[2]));
+    /* go into self test mode 1 */
+    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
     if (rc) {
         return rc;
     }
 
-    /* set self test mode to positive self test */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
+    /* wait 100ms */
+    os_time_delay(OS_TICKS_PER_SEC / 100);
+    rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
     if (rc) {
         return rc;
     }
 
-    os_time_delay(OS_TICKS_PER_SEC / 10);
+    /* take positive offset reading */
+    for(i=0; i<5; i++) {
 
-    /* take self test reading */
-    rc = lis2dw12_get_data(itf, fs, &(pos[0]), &(pos[1]), &(pos[2]));
+        rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
+        if (rc) {
+            return rc;
+        }
+        diff[0] += data[0];
+        diff[1] += data[1];
+        diff[2] += data[2];
+        /* wait at least 20 ms */
+        os_time_delay(OS_TICKS_PER_SEC / 50 + 1);
+    }
+
+    /* go into self test mode 2 */
+    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE2);
     if (rc) {
         return rc;
     }
 
-    /* disable self test mod */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_DISABLE);
+    os_time_delay(OS_TICKS_PER_SEC / 50 + 1);
+    rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
     if (rc) {
         return rc;
     }
 
-    /* calculate accel data difference */
-    change = 0;
-    for(i = 0; i < 3; i++) {
-        change += pos[i] - base[i];
-    }
+    /* take negative offset reading */
+    for (i=0; i<5; i++) {
 
-    if ((change > 70) && (change < 1500)) {
-        *result = 0;
-    } else {
-        *result = -1;
+            rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
+            if (rc) {
+                return rc;
+            }
+            diff[0] -= data[0];
+            diff[1] -= data[1];
+            diff[2] -= data[2];
+            /* wait at least 20 ms */
+            os_time_delay(OS_TICKS_PER_SEC / 50 + 1);
+        }
+
+    /* disable self test mod */
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG1, prev_config, 6);
+        if (rc) {
+            return rc;
+        }
+
+    /* compare values to thresholds */
+    *result = 0;
+    for (i = 0; i < 3; i++) {
+        if ((diff[i] < min) || (diff[i] > max)) {
```

Comment:
My device fails here on negative values, and the st driver you linked does do an abs

Reply:
(gdb) p diff   
$5 = {-3419, -3419, -3419}

(gdb) p diff   
$10 = {-1191, -1878, -1882}

And with an abs Im still failing low occasionally, so Im not sure on the methadology.

(gdb) p diff
$10 = {63, 63, 63}